### PR TITLE
[Synapse  Control Plane] Add vnet features

### PR DESCRIPTION
**Pull request by aim-for-better**
1. Add two properties to existing workspace.json
2. Add new APi privateEndpointConnections.json and privateLinkResources.json
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      "type": "oauth2",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      },
+      "description": "Azure Active Directory OAuth2 Flow"
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Synapse/workspaces/{workspaceName}/privateLinkResources": {
+      "get": {
```
Comment (pilor) [inline specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: >"get" [](start = 6, length = 5)

missing "x-ms-pageable"

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json [hunk 2]
```diff
@@ -779,6 +779,18 @@
             "type": "string"
           },
           "type": "object"
+        },
+        "managedVirtualNetwork": {
+          "description": "Workspace managed virtual network",
+          "type": "string"
+        },
```
Comment (pilor) [inline specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: what is this?  I haven't seen this in other private link reviews

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        "operationId": "PrivateLinkResources_List",
```
Comment (pilor) [inline specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: >List [](start = 45, length = 4)

there should also be an operation to get an individual privateLinkResource by name

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        },
+        "x-ms-examples": {
+          "Delete private endpoint connection": {
+            "$ref": "./examples/DeletePrivateEndpointConnection.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Synapse/workspaces/{workspaceName}/privateEndpointConnections": {
+      "get": {
```
Comment (pilor) [inline specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json]: >"get" [](start = 6, length = 5)

missing "x-ms-pageable"

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+      "x-ms-azure-resource": true
+    },
+    "PrivateLinkServiceConnectionState": {
+      "description": "Connection state details of the private endpoint",
+      "type": "object",
+      "properties": {
+        "status": {
+          "description": "The private link service connection status.",
+          "type": "string"
+        },
```
Comment (pilor) [inline specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json]: this should be a string enum with the possible values

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json [hunk 6]
```diff
@@ -779,6 +779,18 @@
...
+        "privateEndpointConnections": {
+          "description": "Private endpoint connections to the workspace",
+          "allOf": [
+            {
+              "$ref": "./privateEndpointConnections.json#/definitions/PrivateEndpointConnectionList"
+            }
+          ]
```
Comment (pilor) [inline specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: this should be an array of PrivateEndpointConnections, not the List response

**Review by pilor — CHANGES_REQUESTED** [review]
:clock1:

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: > what is this? I haven't seen this in other private link reviews

This property is returned when creating workspace. Please see the ./examples/CreateOrUpdateWorkspace.json

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json (hunk 3 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: > > List [](start = 45, length = 4)
> 
> there should also be an operation to get an individual privateLinkResource by name

Backend doesn't provide such api, I think that list all privateLinkResource is ok.

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json (hunk 1 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: Added

---
In reply to: [397381224](https://github.com/Azure/azure-rest-api-specs/pull/8752#discussion_r397381224) [](ancestors = 397381224)

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json (hunk 4 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json]: > > "get" [](start = 6, length = 5)
> 
> missing "x-ms-pageable"

added

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 6 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: > this should be an array of PrivateEndpointConnections, not the List response

fixed

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json (hunk 3 above)
Comment (pilor) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: A collection GET that exposes entities with resource IDs needs individual GETs for those resources as well.  Some RPs have gotten away without doing this but the majority have done it correctly (search through the repo for examples).

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (pilor) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: That doesn't answer the question about what the property means.  Is it read only?  What values can it have?  What does it do?  The description and other context in the swagger spec does not provide adequate information.

**Review by pilor — CHANGES_REQUESTED** [review]
Some open issues

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: > That doesn't answer the question about what the property means. Is it read only? What values can it have? What does it do? The description and other context in the swagger spec does not provide adequate information.

This is a readonly property mandated by NRP for a workspace that supports private endpoints.  The value can be "default".

**Comment by aim-for-better** [comment]
> Some open issues

Yeah, I an waiting for backend will update it soon

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json (hunk 5 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateEndpointConnections.json]: > this should be a string enum with the possible values

updated

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json (hunk 3 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: > A collection GET that exposes entities with resource IDs needs individual GETs for those resources as well. Some RPs have gotten away without doing this but the majority have done it correctly (search through the repo for examples).

Hi @pilor , Thanks for your comment. I have checked with backend team. Now our service doesn't need to support individual GETs. I think that this is not a block issue right?  Can we ignore this？

**Comment by aim-for-better** [comment]
> Some open issues

Hi @pilor I have updated the PR. Thanks for your comment.

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (pilor) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json]: The description should be updated to indicate what effect this property has and when/why someone should be setting it.  The description right now is just a non-camel cased variant of the property name.

Also, are you sure this shouldn't be a resource ID of a virtual network?  How are you referencing a virtual network by name?  VNet names are not unique by themselves.

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json (hunk 3 above)
Comment (pilor) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json]: No, this violates the RPC and must be implemented

**Review by pilor — CHANGES_REQUESTED** [review]
2 open comments

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: > The description should be updated to indicate what effect this property has and when/why someone should be setting it. The description right now is just a non-camel cased variant of the property name.
> 
> Also, are you sure this shouldn't be a resource ID of a virtual network? How are you referencing a virtual network by name? VNet names are not unique by themselves.

We are sure that this shouldn't be a resource ID of a virtual network. Now this is a read only property return by NRP. Now the value can only be default but we will support more value later so we can't add default value here. It should be string.

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json (hunk 3 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/privateLinkResources.json:26]: > No, this violates the RPC and must be implemented. Refer to cosmos DBs implementation for a good example

Added

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: > The description should be updated to indicate what effect this property has and when/why someone should be setting it. The description right now is just a non-camel cased variant of the property name.
> 
> Also, are you sure this shouldn't be a resource ID of a virtual network? How are you referencing a virtual network by name? VNet names are not unique by themselves.

Setting this to 'default' will ensure that all compute for this workspace is in a virtual network managed on behalf of the user

**Comment by aim-for-better** [comment]
> 2 open comments

addressed

File: specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json (hunk 2 above)
Comment (pilor) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: The description for this property still needs to be updated.  If its readonly as you mention above please mark it as readonly in swagger.

Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: > The description for this property still needs to be updated. How does a user know any of the information that you shared in this thread? If its readonly as you mention above please mark it as readonly in swagger.

Hi @pilor , 
"Setting this to 'default' will ensure that all compute for this workspace is in a virtual network managed on behalf of the user" This description is from backend team, can I use it directly?
Or could you please provide a good description? 

I will mark it as readonly but for the description I have no idea. This is not a block issue. I will update it again with the original description from backend team.

Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: > The description for this property still needs to be updated. How does a user know any of the information that you shared in this thread? If its readonly as you mention above please mark it as readonly in swagger.

I will confirm with backend team to confirm whether it is readonly or not again. Sorry

Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: > The description for this property still needs to be updated. How does a user know any of the information that you shared in this thread? If its readonly as you mention above please mark it as readonly in swagger.

Hi, This property is not readonly.  I am so sorry for previous wrong comment.

Comment (aim-for-better) [reply specification/synapse/resource-manager/Microsoft.Synapse/preview/2019-06-01-preview/workspace.json:786]: I have updated the description to "Setting this to 'default' will ensure that all compute for this workspace is in a virtual network managed on behalf of the user."
This description is provided by backend team. I think that it is correct and make sense.
The only question is that maybe you will ask why not add default value here. 
use "default" to indicate what the server does if the value is not provided by customers. But we ask customer to provide this parameter.

**Review by pilor — APPROVED** [review]

**Pull request merged**